File: app/shared/src/commonTest/kotlin/com/tneff/cyppieagents/comm/AgentAddressingTest.kt
package com.tneff.cyppieagents.comm

import com.tneff.cyppieagents.model.Channel
import com.tneff.cyppieagents.model.ChannelKind
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * CYP-884 (OS-D, Compose mirror of web-ts CYP-876 `agentAddressing.test.ts`) — [resolveAgentChannel] is fail-closed:
 * 1 DIRECT spoke → Resolved; >1 → Ambiguous (NEVER silent-first); 0 → Unreachable (NEVER a fabricated channel). Only
 * DIRECT counts — a GROUP/HUB membership is not a DM spoke.
 */
class AgentAddressingTest {

    private fun ch(id: String, kind: ChannelKind, members: List<String>) = Channel(id, id, kind, members)
    private val spokeFE = ch("po-frontend", ChannelKind.DIRECT, listOf("po", "frontend"))
    private val spokeBE = ch("po-backend", ChannelKind.DIRECT, listOf("po", "backend"))
    private val group = ch("team", ChannelKind.GROUP, listOf("po", "frontend", "backend"))
    private val hub = ch("hub", ChannelKind.HUB, listOf("po", "frontend", "backend"))

    @Test
    fun exactlyOneDirectSpoke_resolvesToThatChannel() {
        assertEquals(AgentAddress.Resolved("po-frontend"), resolveAgentChannel("frontend", listOf(spokeFE, spokeBE, group, hub)))
    }

    @Test
    fun noDirectSpoke_isUnreachable_neverFabricated() {
        // A GROUP/HUB membership is NOT a DM spoke. MUT: fall back to a GROUP/HUB (fabricate a DM route) → reds.
        assertEquals(AgentAddress.Unreachable, resolveAgentChannel("frontend", listOf(group, hub)))
        assertEquals(AgentAddress.Unreachable, resolveAgentChannel("nobody", listOf(spokeFE, spokeBE)))
    }

    @Test
    fun moreThanOneDirectSpoke_isAmbiguous_neverSilentFirst() {
        // MUT: return Resolved(spokes[0]) (silent-first pick) → reds. The ambiguity is surfaced, not guessed away.
        val dupA = ch("po-frontend", ChannelKind.DIRECT, listOf("po", "frontend"))
        val dupB = ch("po-frontend-2", ChannelKind.DIRECT, listOf("po", "frontend"))
        assertEquals(AgentAddress.Ambiguous(listOf("po-frontend", "po-frontend-2")), resolveAgentChannel("frontend", listOf(dupA, dupB)))
    }

    @Test
    fun onlyDirectCounts_groupOrHubWithMember_isNotASpoke() {
        assertEquals(AgentAddress.Unreachable, resolveAgentChannel("frontend", listOf(group)))
        assertEquals(AgentAddress.Unreachable, resolveAgentChannel("frontend", listOf(hub)))
    }
}

File: web-ts/src/comm/agentAddressing.test.ts
import { describe, it, expect } from 'vitest'
import { resolveAgentChannel } from './agentAddressing'
import type { Channel } from '../types/generated/contract'

const ch = (id: string, kind: Channel['kind'], members: string[]): Channel => ({ id, name: id, kind, members })

describe('CYP-876 — resolveAgentChannel (fail-closed, render ≠ authority)', () => {
  const spokeFE = ch('po-frontend', 'DIRECT', ['po', 'frontend'])
  const spokeBE = ch('po-backend', 'DIRECT', ['po', 'backend'])
  const group = ch('team', 'GROUP', ['po', 'frontend', 'backend'])
  const hub = ch('hub', 'HUB', ['po', 'frontend', 'backend'])

  it('★ exactly one DIRECT spoke → resolved to that channel', () => {
    expect(resolveAgentChannel('frontend', [spokeFE, spokeBE, group, hub])).toEqual({ kind: 'resolved', channelId: 'po-frontend' })
  })

  it('★ no DIRECT spoke for the target → UNREACHABLE, never a fabricated channel', () => {
    // A GROUP/HUB membership is NOT a DM spoke. MUT: fall back to a GROUP/HUB (fabricate a DM route) → reds.
    expect(resolveAgentChannel('frontend', [group, hub])).toEqual({ kind: 'unreachable' })
    expect(resolveAgentChannel('nobody', [spokeFE, spokeBE])).toEqual({ kind: 'unreachable' })
  })

  it('★ more than one DIRECT spoke with the target → AMBIGUOUS (null+flag), NEVER silent-first', () => {
    // MUT: return resolved with spokes[0] (silent-first pick) → reds. The ambiguity is surfaced, not guessed away.
    const dupA = ch('po-frontend', 'DIRECT', ['po', 'frontend'])
    const dupB = ch('po-frontend-2', 'DIRECT', ['po', 'frontend'])
    const r = resolveAgentChannel('frontend', [dupA, dupB])
    expect(r.kind).toBe('ambiguous')
    if (r.kind === 'ambiguous') expect([...r.channelIds]).toEqual(['po-frontend', 'po-frontend-2'])
  })

  it('only DIRECT counts — a GROUP/HUB with the member is not a spoke', () => {
    expect(resolveAgentChannel('frontend', [group]).kind).toBe('unreachable')
    expect(resolveAgentChannel('frontend', [hub]).kind).toBe('unreachable')
  })
})
